wdl/src/main/scala/wdl/command/ParameterCommandPart.scala

```diff
@@ -49,29 +45,53 @@ case class ParameterCommandPart(attributes: Map[String, String], expression: Wdl
       case Success(v) => v match {
-        case WomOptionalValue(_, opt) => opt.getOrElse(defaultString)
-        case _ => v
+        case WomOptionalValue(_, opt) => opt.getOrElse(defaultString).validNel
+        case _ => v.validNel
       }
-      case Failure(OptionalNotSuppliedException(_)) => defaultString
-      case Failure(f) => throw new UnsupportedOperationException(s"Could not evaluate expression: ${expression.toWomString}", f)
+      case Failure(OptionalNotSuppliedException(_)) => defaultString.validNel
+      case Failure(_) => s"Could not evaluate expression: ${expression.toWomString}".invalidNel
+    }
+
+    // Do the two arguments represent the same WomFile path.
+    def representTheSamePath(womFile: WomFile, maybeWomFile: WomValue): Boolean = {
+      Option(maybeWomFile) collect { case isWomFile: WomFile if isWomFile.value == womFile.value => true } isDefined
     }
 
-    valueMapper(value) match {
-      case b: WomBoolean if attributes.contains("true") && attributes.contains("false") => if (b.value) attributes.get("true").head else attributes.get("false").head
-      case p: WomPrimitive => p.valueString
-      case a: WomArray if attributes.contains("sep") => a.value.map(_.valueString).mkString(attributes.get("sep").head)
-      case _: WomArray => throw new UnsupportedOperationException(s"Expression '${expression.toString}' evaluated to an Array but no 'sep' was specified")
-      case _ => throw new UnsupportedOperationException(s"Could not string-ify value: $value")
+    // Create the stringified version of the command and record any file created by `writeFile`.
+    def instantiateCommand(value: WomValue): ErrorOr[InstantiatedCommand] = {
+      valueMapper(value) match {
+        case b: WomBoolean if attributes.contains("true") && attributes.contains("false") =>
+          InstantiatedCommand(if (b.value) attributes.get("true").head else attributes.get("false").head).validNel
+        case f: WomFile if !representTheSamePath(f, value) =>
+          // Files generated by writeFiles have "engine-relative" paths which will be different from the container paths
+          // calculated by `valueMapper`. "engine-relative" may mean either the non-Docker container path on the host
+          // running Cromwell, or a cloud path. Capture these newly created files and their engine paths here.
+          val enginePathFile = value.asInstanceOf[WomFile]
+          InstantiatedCommand(f.valueString, List(enginePathFile)).validNel
+        case f: WomFile =>
+          InstantiatedCommand(f.valueString).validNel
+        case p: WomPrimitive => InstantiatedCommand(p.valueString).validNel
+        case a: WomArray if attributes.contains("sep") =>
+          InstantiatedCommand(a.value.map(_.valueString).mkString(attributes.get("sep").head)).validNel
+        case _: WomArray =>
+          s"Expression '${expression.toString}' evaluated to an Array but no 'sep' was specified".invalidNel
```

Comment:
I've seen `${expression.toWomString}` used elsewhere. If that's not the case here, then this should just be `$expression`.

Reply:
This was just massaging the code on the old line 65 to fit the new return type. I'll change this to `${expression.toWomString}` as I suspect that will produce something far more readable. 😄